File: Java/src/Solution101_200/Solution101.java
package Solution101_200;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution101{
    public boolean isSymmetric(TreeNode root) {
        if (root == null){
            return true;
        }
        return dfs(root.left, root.right);
    }

    private boolean dfs(TreeNode left, TreeNode right){
        if (left == null || right == null){
            if (left == null && right == null){
                return true;
            }
            return false;
        }
        if (left.val != right.val){
            return false;
        }
        return dfs(left.left, right.right) && dfs(left.right, right.left);
    }

    public static void main(String[] args) {
        Solution101 solution101 = new Solution101();

    }
}

File: C++/src/Solution101_200/Solution101.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution101.h"

bool Solution101::isSymmetric(TreeNode *root) {
    if (root == nullptr){
        return true;
    }
    return dfs(root->left, root->right);
}

bool Solution101::dfs(TreeNode *left_t, TreeNode *right_t){
    if (left_t == nullptr || right_t == nullptr){
        if (left_t == nullptr && right_t == nullptr){
            return true;
        }
        return false;
    }
    if (left_t->val == right_t->val){
        bool left =  dfs(left_t->left, right_t-> right);
        bool right = dfs(left_t-> right, right_t-> left);
        return left && right;
    }
    return false;
}
